Uniswap V3 native TWAP

kevincharm opened the pull request:
## Summary of work

- Work was started on top of the code pulled from [Kovan](https://kovan.etherscan.io/address/0x568337E3F2D3E773e771a8e81598959DD46D830A#code).
This code seemed to be a very minimal integration with Uniswap V3 and did not take advantage of the fact that all Uniswap V3 pools are now price oracles (they can return TWAPs).
- This new UAV no longer records and keeps observations, but instead queries the respective Uniswap V3 pool's `observe` function to get a TWAP with the set anchor period when it's needed (i.e. when a reporter calls `validate` on the UAV).
- The price of an asset in Uniswap V3 is a function of the "tick" of the pool. The formula is `price = 1.0001^tick`. The math required for conversion between posted prices, comparison to the Uniswap V3 TWAP, and the UAV representation has been modified to support this. Relevant libraries (`TickMath` and `FullMath`) have been included from the Uniswap V3 codebase.
- The tests from the original UAV have been adapted to the new Uniswap V3 architecture - i.e. observations/TWAP tests were removed. Hardhat (+typechain) has been implemented so real Uniswap V3 pools can be used via forked mainnet in tests, and eliminates the need for mocking Uniswap V3 pools.

## TODO

- [x] Move deployer stuff into this repo
- [x] Code coverage (underreporting b/c of incompatibility with resetting fork)
- [x] Strip out old stuff (saddle configs, old tests, Docker stuff)
- [x] Cleanup contracts dir structure
- [x] Autoformat all the things with prettier
- [ ] Update SDK (not sure if this is being used/requires changes)

kevincharm committed:
add code pulled from etherscan

kevincharm committed:
use native Uniswap V3 TWAP

kevincharm committed:
setup hardhat and typescript

kevincharm committed:
fix getUniswapTwap math

kevincharm committed:
preliminarily adapt basic tests to UniV3

kevincharm commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -228,103 +198,66 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
+            // e.g. We want the ETH/USDC price but the actual pool is USDC/ETH.
+            // The TWAP is also down-scaled from Q96 in this step.
+            // -> twap = expScale / (twapX96 / (2**96))
+            twap = mul(expScale, 2**96) / twapX96;
```
It's also possible to avoid this arithmetic (for reversed token pairs) by taking the negative of the time-weighted average tick because

![](https://user-images.githubusercontent.com/10385659/132407581-7b5f836f-8ece-4cd6-914f-d6e6a577aaef.png)
where `p` is price and `t` is tick.

i.e. we can do this instead

```solidity
if (config.isUniswapReversed) {
    timeWeightedAverageTick = -timeWeightedAverageTick;
}
```

alexroan replied:
This division is dangerous. Could we use SafeMath.div() instead?

kevincharm replied:
Definitely in favour of using SafeMath ops - I had taken the approach of modifying as little as possible from the previous code (if not for readability in diffing).

kevincharm committed:
begin migrating UniswapAnchoredView tests

kevincharm committed:
complete UAV#validate tests

kevincharm committed:
implement UAV#getUnderlyingPrice tests

kevincharm committed:
implement UAV#constructor tests

kevincharm committed:
emit FailoverUpdated event on failover (de-)activation

kevincharm committed:
use fake accounts for forked mainnet tests

kevincharm committed:
implement UAV failover tests

kevincharm committed:
implement UAV deactivate failover tests

kevincharm committed:
add exponentiation utility function

kevincharm committed:
implement UniswapConfig tests

kevincharm commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -39,37 +37,20 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
     /**
      * @notice Construct a uniswap anchored view for a set of token configurations
      * @dev Note that to avoid immature TWAPs, the system must run for at least a single anchorPeriod before using.
-     *      NOTE: Reported prices are set to 1 during construction. We assume that this contract will not be voted in by
```
I'm not sure why so many comments were removed in the Kovan deployed version - I'm going to add them back in.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -90,15 +71,9 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
             if (config.priceSource == PriceSource.REPORTER) {
                 require(uniswapMarket != address(0), "reported prices must have an anchor");
-                require(config.reporter != address(0), "reported price must have a reporter");
+                require(config.reporter != address(0), "reported price much have a reporter");
```
Typo?

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -10,7 +10,7 @@ import "../Chainlink/AggregatorValidatorInterface.sol";
 struct Observation {
     uint timestamp;
-    uint acc;
+    int24 acc;
```
Reason for this change?

kevincharm replied:
I've just removed it, since we are no longer recording observations/keeping our own TWAP. This was leftover from the code on Kovan - the accumulative ticks were being recorded as observations, instead of prices.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -135,16 +108,14 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-         // For example, the baseUnit of ETH is 1e18.
-         // Since the prices in this view have 6 decimals, we must scale them by 1e(36 - 6)/baseUnit
+         // Comptroller needs prices in the format: ${raw price} * 1e(36 - baseUnit)
+         // Since the prices in this view have 6 decimals, we must scale them by 1e(36 - 6 - baseUnit)
```
Can we revert this to using the old comment?

We changed because the math in it was slightly off.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -228,103 +198,71 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
+        secondsAgos[1] = 0;
+        (int56[] memory tickCumulatives, ) = IUniswapV3Pool(config.uniswapMarket).observe(secondsAgos);
+
+        int24 timeWeightedAverageTick = int24((tickCumulatives[1] - tickCumulatives[0]) / anchorPeriod_);
```
This seems dangerous. Is there a reason we're using raw arithmetic rather than some form of SafeMath?

Can the tick be negative? How does the tick being positive or negative affect the price?

kevincharm replied:
Yes - these are the minimum and maximum tick values for a Uni V3 Pool: https://github.com/Uniswap/uniswap-v3-core/blob/main/contracts/libraries/TickMath.sol#L9
Price is calculated as a function of the tick as the exponent, like `price = 1.0001^{tick}`. So increasingly negative ticks means price (relative to the other token in the pool) tends towards zero.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -228,103 +198,71 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
         } else {
-            return cumulativePrice0;
+            // Scale up to a common precision (expScale), then down-scale from Q96.
+            twap = mul(expScale, twapX96) / (2**96);
```
Same dangerous div here.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -228,103 +198,71 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-        anchorPrice = mul(unscaledPriceMantissa, config.baseUnit) / ethBaseUnit / expScale;
-
-        emit AnchorPriceUpdated(symbolHash, anchorPrice, oldTimestamp, block.timestamp);
+        uint anchorPrice = mul(unscaledPriceMantissa, config.baseUnit) / ethBaseUnit / expScale;
```
Risky div again.

alexroan commented on package.json:
```diff
@@ -13,16 +13,39 @@
+    "typechain-common-abi": "^0.1.2",
+    "typescript": "^4.4.2"
   },
   "scripts": {
```
There are a few issues we have with the deployer where we've added scripts to help the deployment process.

For example, the Etherscan API limits the amount of parameter data that can be sent to it to verify a contract. To get around this, we created a helper script that outputs flattened contract code compatible with Etherscan UI for verification. We should definitely add that, and other helper scripts here.

kevincharm replied:
Scripts have now been imported from the deployer repo!

kevincharm replied:
The deployment config in `configuration/parameters.js` has been updated with the new V3 pool addresses (`uniswapMarket`). I used the pool with the highest TVL in the case where there's more than one pool (different fees). The REPv2 (Augur) pool seems to have very low liquidity and I've marked it with a comment.

alexroan commented on package.json:
```diff
@@ -13,16 +13,39 @@
     "console": "npx -n --experimental-repl-await saddle console",
-    "compile": "npx saddle compile",
+    "compile": "yarn hardhat compile",
     "deploy": "npx saddle deploy"
```
Needs changing?

alexroan reviewed (changes requested):
Surface level review with a few questions.

Looks like the upgrade to V3 is slimming the code down a lot. Have you tried increasing the number of supported tokens in `UniswapConfig` to see how many more we can add before hitting the contract bytecode size limit?

kevincharm committed:
use reverse tick for isUniswapReversed to reduce arithmetic

kevincharm committed:
remove Observation struct

kevincharm committed:
revert getUnderlyingPrice comment

kevincharm committed:
fix typo in revert message

kevincharm committed:
restore missing comment lines in UAV constructor

kevincharm committed:
use SafeMath for math ops in UAV

* SafeMath lib increases contract size to limit
* optimizer runs changed from 200 -> 100 to fit UAV inside 24KB limit

kevincharm committed:
add scripts/configs from uav-deployer, remove old scripts/configs

kevincharm committed:
remove Dockerfile

kevincharm committed:
move test-specific contracts to contracts/test dir

kevincharm committed:
update uniswapMarket configs with V3 pools

kevincharm committed:
add coverage for contract tests

* solidity-coverage does not support hardhat_reset so coverage is underreported right now
* some tests in PostRealWorldPricesTest were cleaned up

kevincharm committed:
rename PostRealWorldPricesTest.test.ts -> PostRealWorldPrices.test.ts

kevincharm commented:
Thanks @alexroan. I've addressed your comments & questions.

Risky muls & divs have been replaced with `SafeMath` operations, with the exception of:

* Calculating the time-weighted average *tick*, since the tick is a signed int and negative values are valid. I've added a `require` to guard against div-by-zero on `anchorPeriod_`: https://github.com/smartcontractkit/open-oracle/pull/3/files#diff-910d29ce5f6e4bd4a6c5e529ee78206138475f095f7b8093d1566e4f755f9b6cR212
* Calculating the price from the sqrt of the price in Q96 fixed precision (involves squaring a `uint160`). For this, I'm using `mulDiv` from Uniswap's `FullMath` library as it handles 512-bit multiplication: https://github.com/smartcontractkit/open-oracle/pull/3/files#diff-910d29ce5f6e4bd4a6c5e529ee78206138475f095f7b8093d1566e4f755f9b6cR224

Solidity test coverage has been added too, but unfortunately underreports the coverage as it doesn't count tests that reset the fork.

> Looks like the upgrade to V3 is slimming the code down a lot. Have you tried increasing the number of supported tokens in `UniswapConfig` to see how many more we can add before hitting the contract bytecode size limit?

Unfortunately, the addition of `SafeMath` and the extra require above bumped the contract bytecode size up to 24KB so we are now hitting the limit!

kevincharm commented on configuration/parameters.js:
```diff
@@ -0,0 +1,241 @@
+      baseUnit: "1000000000000000000",
+      priceSource: "2",
+      fixedPrice: "0",
+      uniswapMarket: "0xb055103b7633b61518cd806d95beeb2d4cd217e7",
```
https://info.uniswap.org/#/pools/0xb055103b7633b61518cd806d95beeb2d4cd217e7

alexroan replied:
Seems dangerous to push this one through if REP is still in production on Compound, (I don't think it actually is, but I could be wrong).

We should communicate this with them and ask if we can remove altogether.

kevincharm replied:
Indeed, I don't see it on their interface.

se3000 replied:
Can/should we remove it then?

kevincharm replied:
We raised it with Compound folks. They said the market has been deprecated and at some point the protocol hopes to eject it from the system, but they haven't gotten to that point yet.

kevincharm committed:
disable fork reset in coverage, add tests to cover missing branches in UniswapAnchoredView

alexroan commented on .env.example:
```diff
@@ -0,0 +1,4 @@
+COMP_MULTISIG=0xbbf3f1421D886E9b2c5D716B5192aC998af2012c
```
We should add instructions as to how to use this in the README

alexroan commented on .solcover.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  skipFiles: [
+    "test/MockChainlinkOCRAggregator.sol",
```
I've found that adding test contracts to `skipfiles` greatly reduces the coverage in our core repo. What happens when you don't skipe these?

kevincharm replied:
with `skipFiles`:

```
-----------------------------------|----------|----------|----------|----------|----------------|
File                               |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------------------|----------|----------|----------|----------|----------------|
 contracts/                        |       50 |       50 |    57.14 |       50 |                |
  Ownable.sol                      |    18.18 |    33.33 |       40 |       25 |... 38,39,41,48 |
  SafeMath.sol                     |      100 |    66.67 |      100 |      100 |                |
 contracts/Chainlink/              |      100 |      100 |      100 |      100 |                |
  AggregatorValidatorInterface.sol |      100 |      100 |      100 |      100 |                |
 contracts/Uniswap/                |    97.32 |    94.13 |      100 |    95.82 |                |
  UniswapAnchoredView.sol          |    96.39 |    73.91 |      100 |    96.25 |    136,137,138 |
  UniswapConfig.sol                |      100 |    99.24 |      100 |      100 |                |
  UniswapLib.sol                   |    75.81 |    85.42 |      100 |    70.59 |... 117,188,190 |
-----------------------------------|----------|----------|----------|----------|----------------|
All files                          |    96.09 |     92.7 |    91.18 |    94.37 |                |
-----------------------------------|----------|----------|----------|----------|----------------|
```
Note that this still underreports a little: there's a branch in `UniswapAnchoredView.sol` that is definitely covered when running an isolated test, but is reported as not covered here.

without:
```
-----------------------------------|----------|----------|----------|----------|----------------|
File                               |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------------------|----------|----------|----------|----------|----------------|
 contracts/                        |       50 |       50 |    57.14 |       50 |                |
  Ownable.sol                      |    18.18 |    33.33 |       40 |       25 |... 38,39,41,48 |
  SafeMath.sol                     |      100 |    66.67 |      100 |      100 |                |
 contracts/Chainlink/              |      100 |      100 |      100 |      100 |                |
  AggregatorValidatorInterface.sol |      100 |      100 |      100 |      100 |                |
 contracts/Uniswap/                |    94.05 |    91.62 |    81.48 |    91.64 |                |
  UniswapAnchoredView.sol          |    86.75 |    67.39 |    84.62 |    83.75 |... 145,156,157 |
  UniswapConfig.sol                |     98.1 |    98.86 |    91.67 |    97.51 |... 313,314,315 |
  UniswapLib.sol                   |    69.35 |       75 |        0 |    66.18 |... 155,156,157 |
 contracts/test/                   |        0 |        0 |        0 |        0 |                |
  MockChainlinkOCRAggregator.sol   |        0 |      100 |        0 |        0 |          11,15 |
  UniswapV3SwapHelper.sol          |        0 |        0 |        0 |        0 |... 29,31,32,38 |
-----------------------------------|----------|----------|----------|----------|----------------|
All files                          |    91.83 |    89.78 |    68.42 |    89.06 |                |
-----------------------------------|----------|----------|----------|----------|----------------|
```
So it seems to make sense here.

alexroan commented on contracts/SafeMath.sol:
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.6.0 <0.8.0;
```
We should lock this down to the version used elsewhere `^0.6.12`

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -39,29 +35,14 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-
-    /// @notice The event emitted when failover is deactivated
-    event FailoverDeactivated(bytes32 indexed symbolHash);
+    event FailoverUpdated(bytes32 indexed symbolHash, bool status);
```
Two specific functions and events are preferable for this use case IMO. Clarity counts a lot when working with DAOs, as technical and non-technical members will vote on this.

I'd rather we revert to using two separate events and functions as long as it doesn't affect the bytecode size too much.

kevincharm replied:
I agree - this was already here from the Kovan code. Re-adding the separate events now bumps the bytecode size to 24.14KB. How important do you think this would be vs maximising the number of token configs?

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -135,16 +108,16 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
     /**
      * @notice This is called by the reporter whenever a new price is posted on-chain
-     * @dev called by AccessControlledOffchainAggregator
+     * @dev called by AccessControlledOffChainAggregator
```
This is actually lowercase `c`

kevincharm replied:
Also from Kovan - there seems to be quite a few random, seemingly-irrelevant changes 😂

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -153,8 +126,7 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
 
-        // NOTE: We don't do any access control on msg.sender here. The access control is done in getTokenConfigByReporter,
-        // which will REVERT if an unauthorized address is passed.
+        // This will revert if no configs found for the msg.sender
```
Let's revert this comment, as it was a request from the auditor the first time around.

alexroan commented on contracts/Uniswap/UniswapAnchoredView.sol:
```diff
@@ -214,124 +186,82 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
-     * @notice Activate failover, and fall back to using failover directly.
-     * @dev Only the owner can call this function
-     */
-    function activateFailover(bytes32 symbolHash) external onlyOwner() {
```
Same as event comments above ^

alexroan commented on test/UniswapConfig.test.ts:
```diff
@@ -0,0 +1,250 @@
+import { address, keccak256, uint } from "./utils";
+import { smock } from "@defi-wonderland/smock";
+
+use(solidity);
```
Out of interest, what does `use(solidity)` do?

kevincharm replied:
It's what gives the nice matchers like `.to.be.revertedWith` but I just realised Hardhat already provides that automatically without having to explicitly import this 😅 Removed.

alexroan commented on package.json:
```diff
@@ -6,24 +6,47 @@
+    "deploy-test": "yarn hardhat run ./scripts/deploy-uav.js --network hardhat",
+    "verify": "yarn hardhat verify --network mainnet --constructor-args ./configuration/parameters.js",
+    "verify-manual": "yarn hardhat manual-verify",
+    "transfer": "yarn hardhat transfer-ownership --network mainnet --uav"
```
We've got `hardhat-contract-sizer` installed, let's add a script to run it here.

That might require adding `paths` config to hardhat.config.ts.

alexroan reviewed (changes requested):
Nice work so far. I've left comments on the lines here, I've also got some other notes that are worth bringing up:
- Nice use of smock!
- You mentioned that SafeMath has increased the bytecode size of the contract. If we bump the solidity version to `0.8.6` and remove SafeMath, we should definitely see some space clear up. (I think you mentioned this not working. It's definitely worth further investigation). v0.8 will also enable us to remove the experimental encoder.
- On the note of space saving, we might be able to remove some of the variables in `UniswapConfig`. We set every token config up with a `cToken`. The interesting thing about this is that cTokens are upgraded quite often and they don't _have_ to match with the configs here. In that case, when `getTokenConfigByCToken` is called in `UniswapConfig`, it defaults to calling `underlying()` in the unrecognised cToken, and gets the correct config that way. I think we can strip out the `cToken` variable and `getTokenConfigByCToken` altogether from `UniswapConfig` and swap all calls to it to  `getTokenConfigByUnderlying`. This might result in slightly different gas costs (depending on the implementation of the actual cToken), but it would save us a lot of space here and allow us to increase capacity.
- I haven't gone through the new math just yet.
- Have you tried increasing the capacity of `UniswapConfig` to find the upper limit? I think that will increase with the comments above 👆

kevincharm committed:
update README

kevincharm committed:
remove old tests

kevincharm committed:
conform SafeMath solc version to ^0.6.12

kevincharm committed:
update comment AccessControlledOffChainAggregator ->  AccessControlledOffchainAggregator

kevincharm committed:
revert to old comment about getTokenConfigByReporter reverting

kevincharm committed:
add size-contracts script

kevincharm committed:
remove explicit solidity chai matchers (already provided by hardhat)

kevincharm committed:
upgrade to solidity 0.8.7

kevincharm committed:
remove CToken variable and getTokenConfigByCToken function from UniswapConfig, update relevant tests

kevincharm committed:
revert to separate activate/deactivate failover functions and events

kevincharm committed:
bump maximum token configs up to 28

kevincharm committed:
implement bytecode size optimisations on UniswapConfig

kevincharm commented:
Thanks again for the review @alexroan!

>     * Nice use of smock!

I love this tool, thanks for the suggestion :)

>     * You mentioned that SafeMath has increased the bytecode size of the contract. If we bump the solidity version to `0.8.6` and remove SafeMath, we should definitely see some space clear up. (I think you mentioned this not working. It's definitely worth further investigation). v0.8 will also enable us to remove the experimental encoder.
>     * On the note of space saving, we might be able to remove some of the variables in `UniswapConfig`. We set every token config up with a `cToken`. The interesting thing about this is that cTokens are upgraded quite often and they don't _have_ to match with the configs here. In that case, when `getTokenConfigByCToken` is called in `UniswapConfig`, it defaults to calling `underlying()` in the unrecognised cToken, and gets the correct config that way. I think we can strip out the `cToken` variable and `getTokenConfigByCToken` altogether from `UniswapConfig` and swap all calls to it to  `getTokenConfigByUnderlying`. This might result in slightly different gas costs (depending on the implementation of the actual cToken), but it would save us a lot of space here and allow us to increase capacity.

Here is another try at bumping everything to `0.8.7` combined with removing cToken config as you suggested: https://github.com/smartcontractkit/open-oracle/compare/feature/14246-uni-v3-native-twap...smartcontractkit:feature/14246-uni-v3-native-twap-solc-8?expand=1

* Upgraded to 0.8.7
* cToken removed from config, and updated usages
* Made further optimisations in contract bytecode size in UniswapConfig: we can now support *35* token configs
* Experimental encoder removed
* `SafeMath` removed
* `updateFailover` and relevant events reverted to two separate events as before
* Fixed all the new stricter casting rules in 0.8 (this may be adding quite some bytecode)
* I also may have improved the safety of the math in the TWAP calcs because of the new stricter rules on casting

kevincharm committed:
add github actions CI config

kevincharm committed:
remove circleci config

kevincharm committed:
pass in required secrets in GH actions workflow

kevincharm committed:
specify ci environment in GH actions workflow

kevincharm committed:
compile before running tests in gh actions workflow

kevincharm committed:
add 60s test timeout

kevincharm committed:
Merge pull request #4 from smartcontractkit/feature/14246-uni-v3-native-twap-solc-8

Uniswap V3 Native TWAP (v0.8 - merge first)

se3000 commented on contracts/SafeMath.sol:
```diff
@@ -0,0 +1,52 @@
+ * Using this library instead of the unchecked operations eliminates an entire
+ * class of bugs, so it's recommended to use it always.
+ */
+library SafeMath {
```
Seems like there's a lot of places we could still use this but don't. Any particular reason why we do/don't use it in different places?

kevincharm replied:
Good catch; we should remove this library. We're using solc >0.8 now which by default reverts on arithmetic errors. So `SafeMath` has been removed as recommended [here](https://docs.soliditylang.org/en/v0.8.7/080-breaking-changes.html).

se3000 commented on contracts/Uniswap/UniswapLib.sol:
```diff
@@ -1,68 +1,291 @@
+pragma solidity ^0.8.7;
 
-// Based on code from https://github.com/Uniswap/uniswap-v2-periphery
+// From: https://github.com/Uniswap/uniswap-v3-core
```
Can we use their [recommended instructions with NPM](https://github.com/Uniswap/v3-core#local-deployment) instead of just referencing the repo? Then we can lock to a version.

kevincharm replied:
I believe auditors do like when dependencies are all extracted out into the repo like this. This was the main reason it was done this way (please correct me if I'm wrong @alexroan). In the end, I also had to make changes to Uniswap's math library to upgrade to solc 0.8 (stricter rules about not being able to negate uint256s), so would have had to do it anyway, at least for the `FullMath` lib.

se3000 commented on contracts/Uniswap/UniswapConfig.sol:
```diff
@@ -270,67 +312,29 @@ contract UniswapConfig {
-    bool internal immutable isUniswapReversed23;
-    bool internal immutable isUniswapReversed24;
+    // Contract bytecode size optimisation:
+    // Each bit i stores a bool, corresponding to the ith config.
```
🔥

se3000 commented on contracts/Uniswap/UniswapConfig.sol:
```diff
@@ -711,33 +777,377 @@ contract UniswapConfig {
+            reporterMultiplier = reporterMultiplier34;
+        }
+
+        return TokenConfig({
```
Does this save contract space?

kevincharm replied:
Indeed it does free up quite a few storage slots! This change had the most impact on saving contract space.

kevincharm committed:
remove SafeMath

kevincharm committed:
use FullMath.mulDiv and Q96 constant in twap calculation

kevincharm committed:
CVF39: reduce get config calls in constructor

kevincharm committed:
A2: remove redundant type(uint).max checks in getTokenConfigBy* functions

kevincharm committed:
A4: sanity check that reporter is 0 for non-reporter configurations

kevincharm committed:
A6: rewrite if/else branch in priceInternal to suppress false compiler warnings

kevincharm committed:
L1: use else ifs to reduce gas usage in getTokenConfig

kevincharm committed:
A7: remove redundant ternary operation

kevincharm committed:
A1: remove unused struct Slot0 code

kevincharm committed:
A4: fix broken tests that used non-zero reporter for FIXED_ price configs

kevincharm committed:
A8: pin pragma to solc 0.8.7

kevincharm committed:
A5: Re-add comments in fetchAnchorPrice and add clarifications on math

kevincharm committed:
C10, C25: change anchorPeriod stored type to uint32

kevincharm committed:
C24: properly document difference of TWATP vs TWAP in getUniswapTwap() function

kevincharm committed:
C28: remove redundant variable in getUniswapTwap()

kevincharm committed:
C38: use uint256 instead of uint64 for isUniswapReversed

kevincharm committed:
C14: use Uniswap's FullMath.mulDiv(...) to avoid intermediate overflows

kevincharm committed:
C22: check priceSource is reporter before fetching eth price

kevincharm committed:
Merge pull request #7 from smartcontractkit/feature/14246-uni-v3-native-twap-fixes-1

V3 Fixes (Part 1)

kevincharm committed:
Merge pull request #8 from smartcontractkit/feature/14246-uni-v3-native-twap-fixes-2

V3 Fixes (Part 2)

aiagentdev-1990 committed:
update gitignore

aiagentdev-1990 committed:
L-03:  verify price source is reporter when activating failover

aiagentdev-1990 committed:
L-04:  check when downcasting timeWeighterAverageTick to int24

aiagentdev-1990 committed:
N-01:  Add visibility for ethHash variable

aiagentdev-1990 committed:
N-02:  Explicitly declare constants

aiagentdev-1990 committed:
N-06:  rename PriceGuarded event

aiagentdev-1990 committed:
N-07: move CErc20 interface from separate file and import it

aiagentdev-1990 committed:
N-09:  Fix typos

aiagentdev-1990 committed:
N-10:  Explicitly use uint256 instead of uint

aiagentdev-1990 committed:
M-01:  Remove use of encodePacked when not needed

aiagentdev-1990 committed:
L-01:  address gas inefficiencies

aiagentdev-1990 committed:
update tests

alexroan committed:
Smaller require messages

aiagentdev-1990 committed:
update tests

aiagentdev-1990 committed:
N-03:  Update contract comments

aiagentdev-1990 committed:
L02:  cache num configs

aiagentdev-1990 committed:
M01:  Remove use of abi encodePacked

aiagentdev-1990 committed:
address PR feedback

aiagentdev-1990 committed:
update string memory to calldata in getTokenConfigBySymbol function

alexroan committed:
Fix gas test

alexroan committed:
Merge pull request #9 from smartcontractkit/feature/14246-uni-v3-native-twap-open-zeppelin-audits

Address audit feedback
